celery/utils/thread_pool.py

```diff
@@ -0,0 +1,93 @@
+# copy from https://github.com/intuited/terminable_thread
+"""Code based on [http://sebulba.wikispaces.com/recipe+thread2].
+`terminable_thread` provides a subclass of `threading.Thread`,
+  adding the facility to raise exceptions in the context of the given thread.
+See the `README` for information on problematic issues, licensing,
+  and other shenanigans.
+    >>> import time
+    >>>
+    >>> def f():
+    ...     try:
+    ...         while True:
+    ...             time.sleep(0.1)
+    ...     finally:
+    ...         print "outta here"
+    ...
+    >>> t = Thread(target = f)
+    >>> t.start()
+    >>> t.isAlive()
+    True
+    >>> t.terminate()
+    >>> t.join()
+    outta here
+    >>> t.isAlive()
+    False
+"""
+# This program is free software.
+# It comes without any warranty, to the extent permitted by applicable law.
+# You can redistribute it and/or modify it
+#   under the terms of the Do What The Fuck You Want To Public License, Version 2,
+#   as published by Sam Hocevar.
+# See http://sam.zoy.org/wtfpl/COPYING for more details.
+# more inforemathion http://tomerfiliba.com/recipes/Thread2/
+
+import threading
+import inspect
+import ctypes
+from unittest import mock
+from concurrent.futures import ThreadPoolExecutor as _T
+
+from celery.exceptions import WorkerTerminate
+
+
+def _async_raise(tid, exctype):
+    """raises the exception, performs cleanup if needed"""
+    if not inspect.isclass(exctype):
+        raise TypeError("Only types can be raised (not instances)")
+    res = ctypes.pythonapi.PyThreadState_SetAsyncExc(ctypes.c_long(tid),
```

Comment:
TODO: Changed in version 3.7: The type of the id parameter changed from long to unsigned long.

Reply:
Please add a condition here and a TODO comment